@torch.jit.script(fn) now is a torch.jit.Function

zdevito opened the pull request:
This changes the @script decorator to directly return a script::Function object, and fixes up the tests to match this change. It also introduces the equivalent emitFunctionHook to emitModuleHook so that our tests still check for jitter even though all script functions are now no longer modules.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19371 [WIP] Trace directly into first-class module form.
* **#19334 [WIP] @torch.jit.script(fn) now is a torch.jit.Function**
* #19333 pybind CompilationUnit and Function directly
* #19217 Make it possible for self.forward to return a ScriptMethod

Differential Revision: [D14974615](https://our.internmc.facebook.com/intern/diff/D14974615)

zdevito committed:
[WIP] @torch.jit.script(fn) now is a torch.jit.Function

zdevito committed:
Update on "[WIP] @torch.jit.script(fn) now is a torch.jit.Function"

zdevito committed:
Update on "[WIP] @torch.jit.script(fn) now is a torch.jit.Function"

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

suo commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -5,8 +5,8 @@ namespace c10 {
 // c10. Sigh...
-Function* ClassType::getMethod(const std::string& name) const {
-  return compilation_unit_->find_function(name).get();
+std::shared_ptr<Function> ClassType::getMethod(const std::string& name) const {
```
I see a lot of places where we're turning Function refs/ptrs into owning pointers. What's the motivation? Seems like the ownership hasn't really changed since a CompilationUnit still owns the actual functions.

zdevito replied:
We are moving to shared ownership of Function objects across compilation units (and other things). Now that you _can_ own a script::Function, I've been updating the ownership incrementally to reflect that. We might need more tweaks once we finish the 'correct' ownership model.

suo commented on torch/jit/__init__.py:
```diff
@@ -697,6 +704,21 @@ def __getattr__(self, attr):
+    def _import(self, src, constants):
+        """ test import logic for single function, use only for testing """
+        src = "op_version_set = 0\n{}".format(src)
+        # HACK: export always adds a self argument
```
we're planning to remove this before committing to the serialization format for the release, right?

zdevito replied:
Yeah, I need to do an update to the python printer so that it uses the first-class function as its source for serialization. Once I do that then this extra self argument business disappears.

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head
